src/System.Private.CoreLib/shared/System/Text/Rune.cs

```diff
@@ -0,0 +1,724 @@
+            if (value.IsAscii)
+            {
+                return UnicodeUtility.IsInRangeInclusive(value.ValueUnsigned, '0', '9');
+            }
+            else
+            {
+                return (GetUnicodeCategoryNonAscii(value) == UnicodeCategory.DecimalDigitNumber);
+            }
+        }
+
+        public static bool IsLetter(Rune value)
+        {
+            if (value.IsAscii)
+            {
+                return (((value.ValueUnsigned - 'A') & ~0x20u) <= (uint)('Z' - 'A')); // [A-Za-z]
+            }
+            else
+            {
+                return IsCategoryLetter(GetUnicodeCategoryNonAscii(value));
+            }
+        }
+
+        public static bool IsLetterOrDigit(Rune value)
+        {
+            if (value.IsAscii)
+            {
+                return ((AsciiCharInfo[value.Value] & IS_LETTER_OR_DIGIT_FLAG) != 0);
+            }
+            else
+            {
+                return IsCategoryLetterOrDecimalDigit(GetUnicodeCategoryNonAscii(value));
+            }
+        }
+
+        public static bool IsLower(Rune value)
+        {
+            if (value.IsAscii)
+            {
+                return UnicodeUtility.IsInRangeInclusive(value.ValueUnsigned, 'a', 'z');
+            }
+            else
+            {
+                return (GetUnicodeCategoryNonAscii(value) == UnicodeCategory.LowercaseLetter);
+            }
+        }
+
+        public static bool IsNumber(Rune value)
+        {
+            if (value.IsAscii)
+            {
+                return UnicodeUtility.IsInRangeInclusive(value.ValueUnsigned, '0', '9');
+            }
+            else
+            {
+                return IsCategoryNumber(GetUnicodeCategoryNonAscii(value));
+            }
+        }
+
+        public static bool IsPunctuation(Rune value)
+        {
+            return IsCategoryPunctuation(GetUnicodeCategory(value));
+        }
+
+        public static bool IsSeparator(Rune value)
+        {
+            return IsCategorySeparator(GetUnicodeCategory(value));
+        }
+
+        public static bool IsSymbol(Rune value)
+        {
+            return IsCategorySymbol(GetUnicodeCategory(value));
+        }
+
+        public static bool IsUpper(Rune value)
+        {
+            if (value.IsAscii)
+            {
+                return UnicodeUtility.IsInRangeInclusive(value.ValueUnsigned, 'A', 'Z');
+            }
+            else
+            {
+                return (GetUnicodeCategoryNonAscii(value) == UnicodeCategory.UppercaseLetter);
+            }
+        }
+
+        public static bool IsWhiteSpace(Rune value)
+        {
+            if (value.IsAscii)
+            {
+                return (AsciiCharInfo[value.Value] & IS_WHITESPACE_FLAG) != 0;
+            }
+
+            // U+0085 is special since it's a whitespace character but is in the Control category
+            // instead of a normal separator category. No other code point outside the ASCII range
+            // has this mismatch.
```

Comment:
If it's in the control category, vs the whitespace category, wouldn't GetUnicodeCategory say "control" and therefore this would return false anyways?

Reply:
Nevermind, I see that we want it to return true, for some reason.